aif360/sklearn/detectors/facts/__init__.py

```diff
@@ -0,0 +1,354 @@
+
+        Args:
+            X (DataFrame): Dataset given as a :class:`pandas.DataFrame`. As in
+                standard scikit-learn convention, it is expected to contain one
+                instance per row and one feature / explanatory variable per
+                column (labels not needed, we already have an ML model).
+
+        Raises:
+            ValueError: `feats_allowed_to_change` and `feats_not_allowed_to_change`
+                cannot be given simultaneously.
+            Exception: when unreachable code is executed.
+
+        Returns:
+            FACTS: Returns self.
+        """
+        if self.categorical_features is None:
+            self.categorical_features = X.select_dtypes(include=["object", "category"]).columns.to_list()
+        all_feats = X.columns.tolist()
+        if self.feats_allowed_to_change is not None and self.feats_not_allowed_to_change is not None:
+            raise ValueError("Please specify only feats_allowed_to_change or feats_not_allowed_to_change, not both.")
+        elif self.feats_allowed_to_change is None and self.feats_not_allowed_to_change is None:
+            feats_not_allowed_to_change = set()
+        elif self.feats_allowed_to_change is not None:
+            feats_not_allowed_to_change = set(all_feats) - set(self.feats_allowed_to_change)
+        elif self.feats_not_allowed_to_change is not None:
+            feats_not_allowed_to_change = set(self.feats_not_allowed_to_change)
+        else:
+            raise Exception("Code should be unreachable.")
+
+        num_features = list(set(X.columns) - set(self.categorical_features))
+        comparators = feature_change_builder(
+            X=X,
+            num_cols=num_features,
+            cate_cols=self.categorical_features,
+            ord_cols=[],
+            feature_weights=self.feature_weights,
+            num_normalization=False,
+        )
+        params = ParameterProxy(featureChanges=comparators)
+        
+        ifthens_coverage_correctness = valid_ifthens(
+            X=X,
+            model=self.clf,
+            sensitive_attribute=self.prot_attr,
+            freqitem_minsupp=self.freq_itemset_min_supp,
+            feats_not_allowed_to_change=list(feats_not_allowed_to_change)
+        )
+
+        rules_by_if = rules2rulesbyif(ifthens_coverage_correctness)
+
+        print("Computing percentages of individuals flipped by any action with cost up to c, for every c", flush=True)
+        self.rules_with_cumulative = cum_corr_costs_all(
+            rulesbyif=rules_by_if,
+            X=X,
+            model=self.clf,
+            sensitive_attribute=self.prot_attr,
+            params=params,
+        )
+        self.rules_by_if = calc_costs(rules_by_if)
+
+        self.dataset = X.copy(deep=True)
+
+        return self
+    
+    def bias_scan(
+        self,
+        metric: str = "equal-effectiveness",
+        viewpoint: str = "macro",
```

Comment:
What does viewpoint mean?

Reply:
It refers to the notions of "macro viewpoint" and "micro viewpoint" defined in section 2.2 of the [paper](https://arxiv.org/pdf/2306.14978.pdf). As a short explanation, consider a set of actions A and a subgroup (cohort / set of individuals) G. Metrics with the macro viewpoint interpretation are constrained to always apply one action from A to the entire G, while metrics with the micro interpretation are allowed to give each individual in G the min-cost action from A which changes the individual's class.

I should mention, though, that not all combinations of `metric` and `viewpoint` are valid, e.g. "Equal Choice for Recourse" only has a macro interpretation. Do you think this should be reflected in the code somehow?